eth-connector/src/lib.rs

```diff
@@ -540,6 +580,87 @@ impl Deposit for EthConnectorContract {
     }
 }
 
+#[near_bindgen]
+impl FeeManagement for EthConnectorContract {
+    fn get_deposit_fee(&self) -> Option<Fee> {
+        self.fee.deposit_fee
+    }
+
+    fn get_withdraw_fee(&self) -> Option<Fee> {
+        self.fee.withdraw_fee
+    }
+
+    fn get_deposit_fee_per_silo(&self, silo: AccountId) -> Option<Fee> {
+        self.fee.deposit_fee_per_silo.get(&silo)
+    }
+
+    fn get_withdraw_fee_per_silo(&self, silo: AccountId) -> Option<Fee> {
+        self.fee.withdraw_fee_per_silo.get(&silo)
+    }
+
+    fn calculate_fee_amount(
+        &self,
+        amount: U128,
+        fee_type: FeeType,
+        silo: Option<AccountId>,
+    ) -> U128 {
+        let Some(fee) = (match fee_type {
+            FeeType::Deposit => silo.and_then(|silo| self.fee.deposit_fee_per_silo.get(&silo)).or(self.fee.deposit_fee),
+            FeeType::Withdraw => silo.and_then(|silo| self.fee.withdraw_fee_per_silo.get(&silo)).or(self.fee.withdraw_fee),
+        }) else { return 0.into() };
+
+        let fee_amount =
+            amount.0.checked_mul(fee.fee_percentage.0).sdk_unwrap() / FEE_DECIMAL_PRECISION;
+
+        let bounded_fee_amount = if fee.lower_bound.map_or(false, |bound| fee_amount < bound.0) {
+            fee.lower_bound.unwrap().0
+        } else if fee.upper_bound.map_or(false, |bound| fee_amount > bound.0) {
+            fee.upper_bound.unwrap().0
+        } else {
+            fee_amount
+        };
+
+        U128(std::cmp::min(bounded_fee_amount, amount.0))
+    }
+
+    fn set_deposit_fee(&mut self, fee: Option<Fee>) {
+        self.connector.assert_owner_access_right().sdk_unwrap();
+        self.fee.deposit_fee = fee;
+    }
+
+    fn set_withdraw_fee(&mut self, fee: Option<Fee>) {
+        self.connector.assert_owner_access_right().sdk_unwrap();
+        self.fee.withdraw_fee = fee;
+    }
+
+    fn set_deposit_fee_per_silo(&mut self, silo: AccountId, fee: Option<Fee>) {
+        self.connector.assert_owner_access_right().sdk_unwrap();
+        if let Some(fee) = fee {
+            self.fee.deposit_fee_per_silo.insert(&silo, &fee);
+        } else {
+            self.fee.deposit_fee_per_silo.remove(&silo);
+        }
+    }
+
+    fn set_withdraw_fee_per_silo(&mut self, silo: AccountId, fee: Option<Fee>) {
+        self.connector.assert_owner_access_right().sdk_unwrap();
+        if let Some(fee) = fee {
+            self.fee.withdraw_fee_per_silo.insert(&silo, &fee);
+        } else {
+            self.fee.withdraw_fee_per_silo.remove(&silo);
+        }
+    }
+
+    fn claim_fee(&mut self, amount: U128, receiver_id: Option<AccountId>) {
+        self.connector.assert_owner_access_right().sdk_unwrap();
+        self.ft_transfer(
+            receiver_id.unwrap_or_else(env::predecessor_account_id),
+            amount,
+            None,
+        );
+    }
```

Comment:
I think it might be better to have a method that optionally allows you to collect fees on some different account. Thus, it won't need the owner access right of the ETH connector to collect the fees.

On the cons, it would require having an extra `ft_transfer()` call for each deposit/withdrawal and thus also consume more gas.

Reply:
136f9b106410f7bc784862b5e74e852f08fc59e7